File: JAVA/27_ShortestDistanceBacktracking/BOJ_03_14003_가장긴증가하는부분수열5.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] arr = new int[n + 1];
        int[] idxArr = new int[n + 1];
        ArrayList<Integer> list = new ArrayList<>();
        list.add(Integer.MIN_VALUE);
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        for (int i = 1; i <= n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
            if (arr[i] > list.get(list.size() - 1)) {
                idxArr[i] = list.size();
                list.add(arr[i]);
            } else {
                int left = 1, right = list.size() - 1, idx = 1;
                while (left <= right) {
                    int mid = (left + right) / 2;
                    if (list.get(mid) >= arr[i]) {
                        idx = mid;
                        right = mid - 1;
                    } else {
                        left = mid + 1;
                    }
                }
                list.set(idx, arr[i]);
                idxArr[i] = idx;
            }
        }
        int max = list.size() - 1;
        StringBuilder sb = new StringBuilder();
        sb.append(max).append("\n");
        Stack<Integer> stack = new Stack<>();
        while (max > 0) {
            if (idxArr[n] == max) {
                stack.push(arr[n--]);
                max--;
            } else n--;
        }
        while (!stack.isEmpty()) {
            sb.append(stack.pop()).append(" ");
        }
        System.out.println(sb);
    }
}

File: C/27_ShortestDistanceBacktracking/BOJ_03_14003_가장긴증가하는부분수열5.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
int stackIdx = -1, * stack;
void push(int value) {
	stack[++stackIdx] = value;
}
int is_empty() {
	if (stackIdx < 0) return 1;
	else return 0;
}
int pop() {
	if (is_empty()) return INT_MIN;
	else return stack[stackIdx--];
}
main() {
	int n, top = -1;
	scanf("%d", &n);
	int* arr = (int*)malloc(sizeof(int) * (n + 1));
	int* idxArr = (int*)malloc(sizeof(int) * (n + 1));
	int* list = (int*)malloc(sizeof(int) * (n + 1));
	list[++top] = INT_MIN;
	for (int i = 1; i <= n; i++) {
		scanf("%d", &arr[i]);
		if (arr[i] > list[top]) {
			idxArr[i] = ++top;
			list[top] = arr[i];
		}
		else {
			int left = 1, right = top, idx = 1;
			while (left <= right) {
				int mid = (right + left) / 2;
				if (list[mid] >= arr[i]) {
					idx = mid;
					right = mid - 1;
				}
				else {
					left = mid + 1;
				}
			}
			list[idx] = arr[i];
			idxArr[i] = idx;
		}
	}
	printf("%d\n", top);
	stack = (int*)malloc(sizeof(int) * top);
	while (top > 0) {
		if (idxArr[n] == top) {
			push(arr[n--]);
			top--;
		}
		else n--;
	}
	while (!is_empty()) {
		printf("%d ", pop());
	}
}